### PR TITLE
Transcoder capabilities discovery and validation

### PR DESCRIPTION
**Pull request by cyberj0g**
**What does this pull request do? Explain your changes. (required)**
This PR is the last part of https://github.com/livepeer/go-livepeer/issues/2123 implementation. It will allow Orchestrator to send tasks to correct Transcoder according to requested capabilities (codecs). Also it adds capability test at startup for Nvidia GPUs (which may support only a subset of capabilities).

This PR includes changes of `install_ffmpeg.sh` discussed [here](https://github.com/livepeer/go-livepeer/pull/2135). Corresponding [LPMS PR](https://github.com/livepeer/lpms/pull/279) needs to be merged first.

**Specific updates (required)**
* add capability test function which is run on node startup
* add logic to select transcoder according to capabilities in O-T mode
* add logic to validate capabilities in OT mode 

**How did you test each of these updates (required)**
Ran tests

**Does this pull request close any open issues?**
https://github.com/livepeer/go-livepeer/issues/2123

**Checklist:**

- [X] Read the [contribution guide](./doc/contributing.md)
- [X] `make` runs successfully
- [X] All tests in `./test.sh` pass
- [n/a] README and other documentation updated
- [n/a] [Pending changelog](./CHANGELOG_PENDING.md) updated

**Additional notes:**
One missing piece is automatic capability (codec) detection from segment. Currently, if segment is encoded with codec, for which T doesn't support decoding - and decoding capability is not explicitly requested, it will bypass capability-based T selection logic and may fail, even if O has Ts which could decode it. Should `go-livepeer` has such segment codec check (at which point?), or it's best to handle on MistServer side?

### CODE REVIEW

**Review by oscar-davids — APPROVED** [review]
LGTM!

**Comment by cyberj0g** [comment]
Non-obvious quirk on multi-GPU system I just realized: there's no GPU-level segment scheduling logic, and capability discovery will provide a list of capabilities from less capable GPU, so if user has a rig with 5 x RTX 3090 and 1 GTX 745 to hook up the monitor, no codecs except H.264 would show up in capabilities, unless `-nvidia` argument excludes latter GPU. It worth documenting.

File: core/transcoder.go [hunk 1]
```diff
@@ -114,41 +115,66 @@ func (nv *NvidiaTranscoder) Transcode(ctx context.Context, md *SegTranscodingMet
...
+				}
+			}
+		}
+		caps = append(caps, c)
+		test_fail:
 	}
-	defer os.Remove(fname)
-	for _, device := range devices {
-		t1 := NewNvidiaTranscoder(device)
-		// "145x1" is the minimal resolution that succeeds on Windows, so use "145x145"
```
Comment (darkdarkdragon) [inline core/transcoder.go:137]: Can we put this comment back?

File: core/transcoder.go [hunk 2]
```diff
@@ -114,41 +115,66 @@ func (nv *NvidiaTranscoder) Transcode(ctx context.Context, md *SegTranscodingMet
...
+			err = ioutil.WriteFile(fname, mp4testSeg, 0644)
+			if err != nil {
+				glog.Errorf("Error writing test segment for capability %d: %s", c, err)
+				continue
+			}
+			defer os.Remove(fname)
+			// check that capability is supported on all devices
+			for _, device := range devices {
+				t1 := NewNvidiaTranscoder(device)
+				// transcode in 3 sessions which is a maximum on consumer Nvidia GPU with official driver
```
Comment (darkdarkdragon) [inline core/transcoder.go]: Our test intentionally used four output renditions - to let use know about problem at the startup time, not when user started to get real work on production node.

File: server/ot_rpc.go [hunk 3]
```diff
@@ -155,7 +156,10 @@ func runTranscode(n *core.LivepeerNode, orchAddr string, httpc *http.Client, not
 	}
 	ctx = clog.AddSeqNo(ctx, uint64(md.Seq))
 	ctx = clog.AddVal(ctx, "taskId", strconv.FormatInt(notify.TaskId, 10))
-
+	if n.Capabilities!=nil && !md.Caps.CompatibleWith(n.Capabilities.ToNetCapabilities()) {
```
Comment (darkdarkdragon) [inline server/ot_rpc.go]: Can we `gofmt` this?

File: core/orchestrator.go [hunk 4]
```diff
@@ -902,19 +905,32 @@ func removeFromRemoteTranscoders(rt *RemoteTranscoder, remoteTranscoders []*Remo
...
 	defer rtm.RTmutex.Unlock()
 
 	checkTranscoders := func(rtm *RemoteTranscoderManager) bool {
 		return len(rtm.remoteTranscoders) > 0
 	}
 
+	findCompatibleTranscoder := func(rtm *RemoteTranscoderManager) int {
+		for i := len(rtm.remoteTranscoders) - 1; i >= 0; i-- {
+			// no capabilities = default capabilities, all transcoders must support them
+			if caps == nil || caps.bitstring.CompatibleWith(rtm.remoteTranscoders[i].capabilities.bitstring) {
```
Comment (darkdarkdragon) [inline core/orchestrator.go:919]: What will happen if old transcoder which does not present capabilities will be connected to that orchestrator?

File: cmd/livepeer/livepeer.go [hunk 5]
```diff
@@ -901,12 +904,11 @@ func main() {
...
 		*httpAddr = defaultAddr(*httpAddr, "", n.GetServiceURI().Port())
 
-		caps := core.DefaultCapabilities()
 		if *sceneClassificationModelPath != "" {
 			// Only enable experimental capabilities if scene classification model is actually loaded
-			caps = append(caps, core.ExperimentalCapabilities()...)
+			transcoderCaps = append(transcoderCaps, core.ExperimentalCapabilities()...)
 		}
-		n.Capabilities = core.NewCapabilities(caps, core.MandatoryCapabilities())
+		n.Capabilities = core.NewCapabilities(transcoderCaps, core.MandatoryOCapabilities())
```
Comment (darkdarkdragon) [inline cmd/livepeer/livepeer.go:917]: That means that standalone O (which does not do transcoding) will advertise only default capabilities?

**Review by darkdarkdragon — CHANGES_REQUESTED** [review]

File: cmd/livepeer/livepeer.go (hunk 5 above)
Comment (cyberj0g) [reply cmd/livepeer/livepeer.go:917]: Yes. Updating of O capabilities when new T connects is not addressed in this PR, as well as B support for such case. Added [comment](https://github.com/livepeer/go-livepeer/issues/2123#issuecomment-1027782218) to issue.

File: core/transcoder.go (hunk 2 above)
Comment (cyberj0g) [reply core/transcoder.go]: Ok, got it, changed it back to 4 sessions. IIRC the error message without patch won't be verbose, something like `Out of memory`. Is there section in docs which covers possible solutions to this on consumer GPUs?

File: core/transcoder.go (hunk 1 above)
Comment (cyberj0g) [reply core/transcoder.go:137]: Moved it to `capabilities.go` where test segment structure is.

File: server/ot_rpc.go (hunk 3 above)
Comment (cyberj0g) [reply server/ot_rpc.go]: Done.

**Review by cyberj0g — COMMENTED** [review]
Thanks for review, all comments addressed.

File: core/orchestrator.go (hunk 4 above)
Comment (cyberj0g) [reply core/orchestrator.go:919]: Good catch, addressed it in `ot_rpc.go`.

File: core/transcoder.go (hunk 2 above)
Comment (darkdarkdragon) [reply core/transcoder.go]: I know, error message is confusing.
> Is there section in docs which covers possible solutions to this on consumer GPUs?

I'm not sure we want to cover this in our docs - patching drivers is in grey zone from legal viewpoint, so we don't want to encourage it.

File: server/ot_rpc.go [hunk 6]
```diff
@@ -155,7 +156,10 @@ func runTranscode(n *core.LivepeerNode, orchAddr string, httpc *http.Client, not
 	}
 	ctx = clog.AddSeqNo(ctx, uint64(md.Seq))
 	ctx = clog.AddVal(ctx, "taskId", strconv.FormatInt(notify.TaskId, 10))
-
+	if n.Capabilities != nil && !md.Caps.CompatibleWith(n.Capabilities.ToNetCapabilities()) {
+		clog.Errorf(ctx, "Requested capabilities for segment are not compatible with this node taskId=%d url=%s err=%q", notify.TaskId, notify.Url, err)
+		return
```
Comment (darkdarkdragon) [inline server/ot_rpc.go:163]: you need to call `sendTranscodeResult` before return.
Also `err` is nil here, so I think we should define new error to return here.

**Review by darkdarkdragon — CHANGES_REQUESTED** [review]

File: server/ot_rpc.go (hunk 6 above)
Comment (cyberj0g) [reply server/ot_rpc.go:163]: Fixed.

**Review by darkdarkdragon — APPROVED** [review]
LGTM after squash/rebase

**Pull request merged**